go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -280,3 +331,28 @@ func nadNeedsUpdate(oldNAD, newNAD *nettypes.NetworkAttachmentDefinition) bool {
 
 	return !reflect.DeepEqual(oldNAD.Spec, newNAD.Spec)
 }
+
+func (nadController *NetAttachDefinitionController) GetActiveNetworkForNamespace(namespace string) (util.NetInfo, error) {
+	if !util.IsNetworkSegmentationSupportEnabled() {
+		return &util.DefaultNetInfo{}, nil
+	}
+	nadController.primaryNetworks.LockKey(namespace)
+	defer nadController.primaryNetworks.UnlockKey(namespace)
+	primaryNetworks, _ := nadController.primaryNetworks.Load(namespace)
```

Comment:
What happens if the mapping of `network name -> NetInfo` of `nadController.primaryNetworks` disagrees with that on `nadController.networks`? No consequence?
Have you considered to keep `nadController.primaryNetworks` just as a map of `namespace -> network names` and then rely and keep in sync `nadController.networks` to get the actual `NetInfo`?

Reply:
so I tried this and its going to make things more complicated. This is because nadController.networks has no protected parallel access. In fact, nadController itself cannot be run with more than one thread right now. Therefore I would need to introduce either a syncmap for nadController.networks or add a mutex to the nadController itself.

Let's say I just took the simple route and turned the primaryNetworks into a regular map, and used a nadController mutex to lock everything. Now threads all over OVNK trying to get the active network are going to be fighting for the global lock and this is going to cause performance issues. So imo it makes more sense to optimize locking.

To optimize locking I need nadController.networks to have a syncmap lockable key. However, now to get the network from primaryNetworks I need to lock both syncmaps (primaryNetworks and nadController.networks) keys. So we could get into a deadlock scenario with order of locking each of these maps. I'd rather take the tradeoff of duplicate netinfos stored rather than risk complicated dead locks by having maps depend on each other.

Perhaps it would make more sense to change the nadController.networks to contain keys for primary and namespace. For now im just going to leave it like it is if you dont object.

Reply:
I don't think that having a mutex that just guards an unimpeded  access to `nadController.networks` and `nadController.primaryNetworks` is going to hurt performance.

I think internally in order to lock a specific key for the sync map, you first need need to lock the map itself, lock the key, then unlock the map. So you are incurring in the same kind of penalty there, if any, with that map lock. 

The sync map just makes sense when you need the key locked for a meaningful amount of time I would say.

If I cannot convince you, then its fine to leave things as they are.

Reply:
I see your point, but consider some other things. Right now nadController does very little work when it processes a NAD so it wont hold the lock for very long, especially since its single threaded. It takes nm.lock and then enqueues a key. However, network manager also takes this nm lock when it syncs an object. This can take a long time when its trying to configure everything in OVN.

So now imagine many UDNs, namespaces, pods/services, network policy being created at the same time. pod handlers have 15 threads. We could now have 10s of threads all trying to GetActiveNetworkForNamespace for all different namespaces. Now, if nm is syncing a namespace z, and all these other threads are trying to GetActiveNetworkForNamespace (namespaces A-Y), they are all blocked.

So the NAD controller/NM thread has single handedly stalled other core controllers in ovnk.

This is a similar problem to what we used to have with the infamous  namespaceInfo mutex. We had a ton of other controllers trying to get information about specific namespaces, but the global lock was blocking all of them. That's when I turned to try to implement lockable keys, and Nadia improved this by making a new syncmap type.

Reply:
I think we wouldn't need for this new lock to guard all the processing of nadController. Just literally the access to the maps and that's it, because NetInfo has its own locks. And that's why I am thinking this critical region is comparable with the critical region you would already have with SyncMap.

BTW, you have a point on the `nm.lock`. Starting a controller can take time (it shouldn't, but unfortunately it might). However, the need of that lock is also only to protect access to `nm.networks` so it could guard only just that. Potential improvement.

Reply:
After our discussion I am posting a new version where we reduce the network tracking to just NetworkManager and remove it from NAD Controller, as well as modify locking to reduce contention on GetActiveNetworkForNamespace.